src/piadm.sh

```diff
@@ -655,13 +675,250 @@ ispoolenabled() {
+	# We put a README (see cat_readme() above) and the string "ipxe"
+	# for the PI-stamp.
+	mkdir -p ./platform-ipxe/etc/version
+	cat_readme > ./platform-ipxe/README
+	echo "ipxe" > ./platform-ipxe/etc/version/platform
+
+	# Now we set up the "etc" directory, which contains versions of
+	# both loader ("boot") and iPXE ("ipxe").
+	mkdir -p etc/version
+	cp -f ${TRITON_IPXE_ETC}/version/* etc/version/.
+	vecho "installing ipxe version: " "$(cat etc/version/ipxe)"
+
+	# Now we set up the "boot-ipxe" directory.
+	mkdir boot-ipxe
+	# Use tar here because it's the first time.
+	tar -cf - -C ${TRITON_IPXE_BOOT} . | tar -xf - -C boot-ipxe
+	# Preserve versions in boot-ipxe too in case we need them later.
+	cp -f etc/version/boot boot-ipxe/bootversion
+	cp -f etc/version/ipxe boot-ipxe/ipxeversion
+
+	# Symlinks for loader default and `piadm list` consistency.
+	ln -s platform-ipxe platform
+	ln -s boot-ipxe boot
+
+	# Install a PI for backup booting purposes.
+	if ! install_pi_CN "$activestamp" && \
+		[[ "$CNAPI_DEFAULT_PI" != "$activestamp" ]]; then
+		/bin/rm -rf platform-"$activestamp"
+		if ! install_pi_CN "$CNAPI_DEFAULT_PI"; then
+			/bin/rm -rf platform-"$CNAPI_DEFAULT_PI"
+			err "No PIs available"
+		fi
+	fi
+	# installstamp will be set by the successful install_pi_CN()
+
+	# Populate loader.conf.
+	# NOTE: One could uncomment the sed below and replace the cp if one
+	# wished to have the CN backup-boot not go into the Triton HN
+	# installer but act in a different kind of weird way.
+	# sed 's/headnode="true"/headnode="false"/g' \ <
+	#	boot-ipxe/loader.conf.tmpl > boot-ipxe/loader.conf
+	cp boot-ipxe/loader.conf.tmpl boot-ipxe/loader.conf
+	{
+		echo 'ipxe="true"'
+		echo 'smt_enabled="true"'
+		echo 'console="ttyb,ttya,ttyc,ttyd,text"'
+		echo 'os_console="ttyb"'
+		echo 'fstype="ufs"'
+		# use $installstamp to help!
+		echo "platform-version=$installstamp"
+		# Need an extra "platform" in here to satisfy the illumos
+		# load-time that looks for a unix in the boot archive.  The
+		# boot file path MUST have /platform/i86pc/kernel/amd64/unix
+		# as its trailing components.  See install_pi_CN() for the
+		# insertion of a symlink to help out.
+		echo bootfile=\"/platform-"$installstamp"/platform/i86pc/kernel/amd64/unix\"
+		echo boot_archive_name=\"/platform-"$installstamp"/i86pc/amd64/boot_archive\"
+		echo boot_archive.hash_name=\"/platform-"$installstamp"/i86pc/amd64/boot_archive.hash\"
+	} >> boot-ipxe/loader.conf
+
+	# Caller will invoke update_boot_sectors.
+}
+
+update_CN() {
+	declare -a pdirs
+
+	if [[ "$TRITON_CN" != "yes" ]]; then
+		err "The update command may only be used on a Triton Compute Node"
+	fi
+
+	piname_present_get_bootfs "ipxe" "$1"
+	cd "/${bootfs}" || fatal "Could not chdir to /$bootfs"
+
+	# First check if the backup PI is in need of update.
+	# The standard iPXE/CN deployment has exactly one platform-STAMP.
+	mapfile -t pdirs < <(ls -d platform-[0-9]*T*Z)
+	if [[ ${#pdirs[@]} -gt 1 ]]; then
```

Comment:
`"${#pdirs[@]}"`

From the Bash guide:
`The syntax "${myfiles[@]}" is extremely important. It works just like "$@" does for the positional parameters: it expands to a list of words, with each array element as one word, no matter what it contains. Even if there are spaces, tabs, newlines, quotation marks, or any other kind of characters in one of the array elements, it'll still be passed along as one word to whichever command we're running. `

Reply:
I had tested this and it worked.  What are you worried about here exactly?  I don't see a number needing to be quoted here.

Reply:
It doesn’t need to be quoted because it’s not expanding the variable to something that can potentially glob. It’s always going to be an integer (length of the array).

Reply:
You know what I completely blew past it was getting the len() of the array.  I missed the `#`.  Ignore me!